File: src/main/java/org/kite9/diagram/visualization/compaction/rect/AbstractRectangularizer.java
package org.kite9.diagram.visualization.compaction.rect;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.kite9.diagram.common.algorithms.so.Slideable;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.common.objects.Rectangle;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.position.Turn;
import org.kite9.diagram.visualization.compaction.AbstractCompactionStep;
import org.kite9.diagram.visualization.compaction.Compaction;
import org.kite9.diagram.visualization.compaction.Compactor;
import org.kite9.diagram.visualization.compaction.segment.Segment;
import org.kite9.diagram.visualization.display.CompleteDisplayer;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.diagram.visualization.orthogonalization.DartFace;
import org.kite9.diagram.visualization.orthogonalization.DartFace.DartDirection;
import org.kite9.framework.logging.LogicException;

/**
 * This class is responsible for 'completing' a dart diagram by ensuring that
 * each face is subdivided by darts into rectangles.
 * 
 * This is done because otherwise the algorithm for compaction cannot work out
 * exactly where to align the darts.
 * 
 * The way this works is by tracing around each face looking for LRR or RRL
 * patterns of angles between the segments in the face.
 * 
 * There are 5 {@link VertexTurn}s in a rectangularization. These are
 * labelled as follows:
 * 
 * <pre>
 *       |          <- (post - not changed)
 *        ------    <- (meets)
 *              |   <-  (link)
 *           ---    <-  (par)
 *          |       <- (extender)
 *          |
 * </pre>
 * 
 * The naming is because the extender is extended to the meets item. par is
 * clearly parallel to meets, while link is linking par to meets.
 * 
 * @author robmoffat
 * 
 */
public abstract class AbstractRectangularizer extends AbstractCompactionStep {

	public AbstractRectangularizer(CompleteDisplayer cd) {
		super(cd);
	}

	boolean failfast = false;
	
	/**
	 * This ties off any loose ends in the diagram by extending the segments to
	 * meet each other. This prevents overlapping of darts in the diagram.
	 * overlapping.
	 */
	@Override
	public void compact(Compaction c, Rectangular r, Compactor rc) {
		List<DartFace> faces = c.getOrthogonalization().getDartFacesForRectangular(r);
		
		if (faces == null) {
			return;
		}
		
		List<Dart> result = new ArrayList<Dart>();
		List<DartFace> orderedFaces = new ArrayList<DartFace>(faces);

		for (DartFace df : orderedFaces) {
			log.send(log.go() ? null : "FACE: " + df);
			// first, add all the segments to the stack in unrectangularized
			// form
			List<VertexTurn> theStack = new ArrayList<VertexTurn>();
			List<DartDirection> turns = df.getDartsInFace();
			Vertex from = df.getStartVertex();
			buildStack(df, theStack, turns, from, c);
			
			// ensures basic dart separation
			for (int i = 0; i < theStack.size(); i++) {
				fixSize(c, getIthElementRotating(theStack, i), 0, !df.outerFace);
			}

			performFaceRectangularization(c, result, theStack);

			if (!df.outerFace) {

				if (theStack.size() != 4) {
					throw new LogicException("Rectangularization did not complete properly - stack > 4, face = "+df);
				}
			}

			for (int i = 0; i < theStack.size(); i++) {
				fixSize(c, getIthElementRotating(theStack, i), 0, !df.outerFace);
			}
			
			setSlideableFaceRectangle(c, df, theStack, df.outerFace);
		}

	}

	private void buildStack(DartFace df, List<VertexTurn> theStack,
			List<DartDirection> turns, Vertex from, Compaction c) {
		if (df.getDartsInFace().size() > 2) {
			
			int startPoint = 0;
			for (int i = 0; i <  turns.size(); i++) {
				DartDirection last = turns.get((turns.size() -1 + i ) % turns.size());
				DartDirection current = turns.get(i);
				if (last.getDirection() != current.getDirection()) {
					startPoint = i;
					break;
				}
			}
			
			List<DartDirection> turnCopy = new ArrayList<>(turns);
			Collections.rotate(turnCopy, -startPoint);
			
			List<Segment> segments = turnCopy.stream().map(dd -> c.getSegmentForDart(dd.getDart())).collect(Collectors.toList());
			List<Direction> directions = turnCopy.stream().map(dd -> dd.getDirection()).collect(Collectors.toList());
					
			List<Segment> uniqueSegments = new ArrayList<>();
			List<Direction> uniqueDirections = new ArrayList<>();
			for (int i = 0; i < segments.size(); i++) {
				Segment segment = segments.get(i);
				Direction d = directions.get(i);
				if ((i == 0) ||  (segment != uniqueSegments.get(uniqueSegments.size()-1))) {
					uniqueSegments.add(segment);
					uniqueDirections.add(d);
				}
			}
			
			// convert to VertexTurns
			int us = uniqueSegments.size();
			for (int i = 0; i < us; i++) {
				
				Segment last = uniqueSegments.get((i-1+us) % us);
				Segment current = uniqueSegments.get(i);
				Segment next = uniqueSegments.get((i+1) % us);
				Direction d = uniqueDirections.get(i);
				
				VertexTurn t = new VertexTurn(c, current.getSlideable(), d, last.getSlideable(), next.getSlideable());
				theStack.add(t);
			}
		
			log.send("Stack for face "+df, theStack);
		}
	}

	protected abstract void performFaceRectangularization(Compaction c, List<Dart> result, List<VertexTurn> theStack);

	private void setSlideableFaceRectangle(Compaction c, DartFace df, List<VertexTurn> theStack, boolean outer) {
		Rectangle<Slideable<Segment>> r = new Rectangle<>(
				getSlideableInDirection(theStack, outer ? Direction.LEFT : Direction.RIGHT),
				getSlideableInDirection(theStack, outer ? Direction.UP : Direction.DOWN),
				getSlideableInDirection(theStack, outer ? Direction.RIGHT : Direction.LEFT), 
				getSlideableInDirection(theStack, outer ? Direction.DOWN : Direction.UP));
			

		c.createFaceSpace(df, r);
	}

	private Slideable<Segment> getSlideableInDirection(List<VertexTurn> vt, Direction d) {
		for (int i = 0; i < vt.size(); i++) {
			VertexTurn prev = vt.get(( i + vt.size() -1 ) % vt.size());
			VertexTurn curr = vt.get(i);
			VertexTurn next = vt.get(( i + 1 ) % vt.size());
			
			if ((curr.getDirection() == d) && (prev.getDirection() != next.getDirection())) {
				return curr.getSlideable();
			}
		}		

		throw new LogicException("No turn in that direction");
	}

	/**
	 * Works out the direction of turn between one segment and the next
	 */
	public static Turn getTurn(VertexTurn t1, VertexTurn t2) {
		Direction thisDirection = t1.getDirection();
		Direction nextDirection = t2.getDirection();

		Turn change = thisDirection.getDirectionChange(nextDirection);
		return change;
	}

	public static <X> X getIthElementRotating(List<X> items, int i) {
		return items.get((i + items.size()) % items.size());
	}

	protected static VertexTurn getItemRotating(List<VertexTurn> stack, int index) {
		while (index < 0)
			index += stack.size();
		index = index % stack.size();
		return stack.get(index);
	}

	protected void performRectangularizationD(List<VertexTurn> stack, Compaction c, List<Dart> out, VertexTurn ext,
			VertexTurn par, VertexTurn link, VertexTurn meets) {
		logRectangularizationContext(ext, par, link, meets);
		Slideable<Segment> first = ext.getEndsWith();
		Slideable<Segment> to = meets.getSlideable();
		Direction d2 = Direction.reverse(meets.getDirection());
		Direction d1 = ext.getDirection();

		performRectangularization(c, out, meets, link, par, ext, first, to, d1, d2);
		cutRectangleCorner(stack, par, link);
	}

	/**
	 * Given that fixing is in a rectangle, with sides of before and after,
	 * there may need to be a minimum length set on fixing.
	 * 
	 * This errs on the side of too large right now
	 */
	protected void fixSize(Compaction c, VertexTurn link, double externalMin, boolean concave) {
		Slideable<Segment> early = link.getStartsWith();
		Slideable<Segment> late = link.getEndsWith();
		Segment early1 = (Segment) early.getUnderlying();
		Segment late1 = (Segment) late.getUnderlying();
		Direction d = link.getDirection();
		
		double minDistance = getMinimumDistance(Direction.isHorizontal(d), early1, late1,  link.getSegment(), concave);
		log.send(log.go() ? null : "Fixing: "+link+" min length "+minDistance);
		link.ensureLength(Math.max(minDistance, externalMin));
	
	}

	protected void performRectangularizationA(List<VertexTurn> stack, Compaction c, List<Dart> out, VertexTurn meets,
			VertexTurn link, VertexTurn par, VertexTurn ext) {
		logRectangularizationContext(meets, link, par, ext);
		Slideable<Segment> first = ext.getStartsWith();
		Slideable<Segment> to = meets.getSlideable();
		Direction d1 = Direction.reverse(ext.getDirection());
		Direction d2 = meets.getDirection();

		performRectangularization(c, out, meets, link, par, ext, first, to, d1, d2);
		cutRectangleCorner(stack, link, par);
	}
	

	protected void performRectangularization(Compaction c, List<Dart> out, VertexTurn meets, VertexTurn link,
			VertexTurn par, VertexTurn extender, Slideable<Segment> from, Slideable<Segment> to, Direction d1, Direction d2) {
		fixSize(c, link, 0, true);
		fixSize(c, par, 0, true);

		double newMeetsLength = calculateNewMeetsLength(meets, par);
		double extensionLength = extender.getMinimumLength() + link.getMinimumLength(); 

		if (extender.getStartsWith() == from) {
			extender.resetEndsWith(to);
		} else {
			extender.resetStartsWith(to);
		}
		
		// update meets
		if (meets.getStartsWith() == link.getSlideable()) {
			meets.resetStartsWith(extender.getSlideable());
		} else {
			meets.resetEndsWith(extender.getSlideable());
		}
		
		fixSize(c, meets, newMeetsLength, true);
		fixSize(c, extender, extensionLength, true);
	}

	protected double calculateNewMeetsLength(VertexTurn meets, VertexTurn par) {
		System.out.println("fix this");
		if (par.isLengthKnown()) {
			return Math.max(0, calculateLength(meets) - calculateLength(par));
		} else {
			return 0;
		}
	}

	private int calculateLength(VertexTurn vt) {
		return vt.getMinimumLength();
	}

	private void logRectangularizationContext(VertexTurn vt4, VertexTurn vt3, VertexTurn vt2, VertexTurn vt1) {
		log.send(log.go() ? null : "Context:");
		log.send(log.go() ? null : "vt4 " + vt4);
		log.send(log.go() ? null : "vt3 " + vt3);
		log.send(log.go() ? null : "vt2 " + vt2);
		log.send(log.go() ? null : "vt1 " + vt1);
	}

	/**
	 * After the new darts are added to create the rectangle, this snips off the
	 * old rectangle from the stack
	 */
	private void cutRectangleCorner(List<VertexTurn> stack, VertexTurn remove1, VertexTurn remove2) {
		stack.remove(remove1);
		stack.remove(remove2);
		log.send(log.go() ? null : "Removed: " + remove1);
		log.send(log.go() ? null : "Removed: " + remove2);
	}

	public static List<Turn> patternA = createList(Turn.LEFT, Turn.RIGHT, Turn.RIGHT);
	public static List<Turn> patternD = createList(Turn.RIGHT, Turn.RIGHT, Turn.LEFT);

	private static List<Turn> createList(Turn... turns) {
		List<Turn> out = new ArrayList<Turn>();
		for (Turn t : turns) {
			out.add(t);
		}
		return out;
	}

	public String getPrefix() {
		return "RECT";
	}

	public boolean isLoggingEnabled() {
		return true;
	}

}

File: src/main/kotlin/org/kite9/diagram/visualization/compaction/rect/AbstractRectangularizer.kt
package org.kite9.diagram.visualization.compaction.rect

import org.kite9.diagram.common.Collections
import org.kite9.diagram.common.algorithms.so.Slideable
import org.kite9.diagram.common.objects.Rectangle
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Turn
import org.kite9.diagram.visualization.compaction.*
import org.kite9.diagram.visualization.compaction.rect.PrioritisedRectOption.TurnShape
import org.kite9.diagram.visualization.compaction.segment.Segment
import org.kite9.diagram.visualization.display.CompleteDisplayer
import org.kite9.diagram.visualization.orthogonalization.DartFace
import org.kite9.diagram.visualization.orthogonalization.DartFace.DartDirection

/**
 * This class is responsible for 'completing' a dart diagram by ensuring that
 * each face is subdivided by darts into rectangles.
 *
 * This is done because otherwise the algorithm for compaction cannot work out
 * exactly where to align the darts.
 *
 * The way this works is by tracing around each face looking for LRR or RRL
 * patterns of angles between the segments in the face.
 *
 * There are 5 [VertexTurn]s in a rectangularization. These are
 * labelled as follows:
 *
 * <pre>
 * |          <- (post - not changed)
 * ------    <- (meets)
 * |   <-  (link)
 * ---    <-  (par)
 * |       <- (extender)
 * |
</pre> *
 *
 * The naming is because the extender is extended to the meets item. par is
 * clearly parallel to meets, while link is linking par to meets.
 *
 * @author robmoffat
 */
abstract class AbstractRectangularizer(cd: CompleteDisplayer) : AbstractCompactionStep(cd) {

    var failfast = false

    /**
     * This ties off any loose ends in the diagram by extending the segments to
     * meet each other. This prevents overlapping of darts in the diagram.
     * overlapping.
     */
    override fun compact(c: Compaction, r: Embedding, rc: Compactor) {
        log.send("Rectangularizing: $r")
        val faces = r.dartFaces ?: return
        val orderedFaces = selectFacesToRectangularize(c, faces)
        log.send("Rectangularizing faces: ", orderedFaces)
        val stacks = setupDartFaceStacks(c, orderedFaces)
        performSecondarySizing(c, stacks)
        performFaceRectangularization(c, stacks)
        for (df in orderedFaces) {
            val theStack = stacks[df]
            if (theStack != null) {
                if (!df!!.outerFace) {
                    if (theStack.size != 4) {
                        throw LogicException("Rectangularization did not complete properly - stack > 4, face = $df")
                    }
                }
                for (i in theStack.indices) {
                    fixSize(c, getIthElementRotating(theStack, i), 0.0, !df.outerFace, false)
                }
                setSlideableFaceRectangle(c, df, theStack, df.outerFace)
            }
        }
    }

    open fun performSecondarySizing(c: Compaction, stacks: Map<DartFace, MutableList<VertexTurn>>) {

    }

    protected abstract fun selectFacesToRectangularize(c: Compaction, faces: List<DartFace>): List<DartFace>

    protected fun setupDartFaceStacks(c: Compaction, orderedFaces: List<DartFace>): MutableMap<DartFace, MutableList<VertexTurn>> {
        val stacks: MutableMap<DartFace, MutableList<VertexTurn>> = HashMap()
        for (df in orderedFaces) {
            log.send(if (log.go()) null else "Creating Face Stack for: $df")
            // first, add all the segments to the stack in unrectangularized
            // form
            val theStack: MutableList<VertexTurn> = ArrayList()
            val turns: List<DartDirection> = df.dartsInFace
            buildStack(df, theStack, turns, c)
            for (i in theStack.indices) {
                fixSize(c, getIthElementRotating(theStack, i), 0.0, isConcave(theStack, i), true)
            }
            stacks[df] = theStack
        }
        return stacks
    }

    private fun isConcave(theStack: List<VertexTurn?>, i: Int): Boolean {
        val prev = getIthElementRotating(theStack, i - 1)
        val next = getIthElementRotating(theStack, i + 1)
        return prev!!.direction !== next!!.direction
    }

    private fun buildStack(
        df: DartFace,
        theStack: MutableList<VertexTurn>,
        turns: List<DartDirection>,
        c: Compaction
    ) {
        if (df.dartsInFace.size > 2) {
            var startPoint = 0
            for (i in turns.indices) {
                val last = turns[(turns.size - 1 + i) % turns.size]
                val current = turns[i]
                if (last.direction !== current.direction) {
                    startPoint = i
                    break
                }
            }
            val turnCopy: MutableList<DartDirection> = ArrayList(turns)
            java.util.Collections.rotate(turnCopy, -startPoint)

            val segments = turnCopy.map { c.getSegmentForDart( it.dart ) }

            val directions = turnCopy.map { it.direction }
            val uniqueSegments: MutableList<Segment> = ArrayList()
            val uniqueDirections: MutableList<Direction> = ArrayList()
            for (i in segments.indices) {
                val segment = segments[i]
                val d = directions[i]
                if (i == 0 || segment != uniqueSegments[uniqueSegments.size - 1]) {
                    uniqueSegments.add(segment)
                    uniqueDirections.add(d)
                }
            }

            // convert to VertexTurns
            val us = uniqueSegments.size
            for (i in 0 until us) {
                val last = uniqueSegments[(i - 1 + us) % us]
                val current = uniqueSegments[i]
                val next = uniqueSegments[(i + 1) % us]
                val d = uniqueDirections[i]
                val t = VertexTurn(i, c, current.slideable!!, d, last.slideable!!, next.slideable!!, df.partOf)
                theStack.add(t)
            }
            log.send("Stack for face $df", theStack)
        }
    }

    protected abstract fun performFaceRectangularization(c: Compaction, stacks: MutableMap<DartFace, MutableList<VertexTurn>>)

    private fun setSlideableFaceRectangle(c: Compaction, df: DartFace, theStack: List<VertexTurn>, outer: Boolean) {
        val r = Rectangle(
            getSlideableInDirection(theStack, if (outer) Direction.LEFT else Direction.RIGHT, outer),
            getSlideableInDirection(theStack, if (outer) Direction.UP else Direction.DOWN, outer),
            getSlideableInDirection(theStack, if (outer) Direction.RIGHT else Direction.LEFT, outer),
            getSlideableInDirection(theStack, if (outer) Direction.DOWN else Direction.UP, outer)
        )
        c.createFaceSpace(df, r)
    }

    private fun getSlideableInDirection(vt: List<VertexTurn>, d: Direction, outer: Boolean): FaceSide {
        val others: MutableSet<Slideable<Segment>> = HashSet()
        var main: Slideable<Segment>? = null
        for (i in vt.indices) {
            val prev = vt[(i + vt.size - 1) % vt.size]
            val curr = vt[i]
            val next = vt[(i + 1) % vt.size]
            if (curr.direction === d) {
                if (prev.direction !== next.direction) {
                    main = curr.slideable
                }
                if (outer == true) {
                    others.add(curr.slideable)
                }
            }
        }
        if (main == null) {
            throw LogicException("No turn in that direction")
        }
        return FaceSide(main, others)
    }

    protected fun performRectangularizationD(
        stack: MutableList<VertexTurn>, c: Compaction, ext: VertexTurn,
        par: VertexTurn, link: VertexTurn, meets: VertexTurn, shape: TurnShape
    ) {
        // logRectangularizationContext(ext, par, link, meets);
        val first: Slideable<Segment> = ext.endsWith
        val to: Slideable<Segment> = meets.slideable
        performRectangularization(c, meets, link, par, ext, first, to, shape)
        cutRectangleCorner(stack, par, link)
    }

    /**
     * Given that fixing is in a rectangle, with sides of before and after,
     * there may need to be a minimum length set on fixing.
     *
     * This errs on the side of too large right now
     * @param initialSetting
     */
    protected fun fixSize(
        c: Compaction?,
        link: VertexTurn?,
        externalMin: Double,
        concave: Boolean,
        initialSetting: Boolean
    ) {
        val early = if (link!!.increasingDirection()) link.startsWith else link.endsWith
        val late = if (link.increasingDirection()) link.endsWith else link.startsWith
        val early1 = early.underlying
        val late1 = late.underlying
        log.send(if (log.go()) null else " Early: $early late: $late")
        val along = if (initialSetting) link.segment else null
        val minDistance = getMinimumDistance(early1, late1, along, concave)
        link.ensureMinLength(Math.max(minDistance, externalMin))
        log.send(if (log.go()) null else "Fixed: $link min length $minDistance")
        return
    }

    protected fun performRectangularizationA(
        stack: MutableList<VertexTurn>, c: Compaction, meets: VertexTurn,
        link: VertexTurn, par: VertexTurn, ext: VertexTurn, shape: TurnShape
    ) {
        // logRectangularizationContext(meets, link, par, ext);
        val first: Slideable<Segment> = ext.startsWith
        val to: Slideable<Segment> = meets.slideable
        performRectangularization(c, meets, link, par, ext, first, to, shape)
        cutRectangleCorner(stack, link, par)
    }

    protected fun performRectangularization(
        c: Compaction, meets: VertexTurn, link: VertexTurn,
        par: VertexTurn, extender: VertexTurn, from: Slideable<Segment>, to: Slideable<Segment>, shape: TurnShape
    ) {
        val newExtenderLength = extender.getLength(false) + link.getLength(false)
        if (extender.startsWith == from) {
            extender.resetEndsWith(to, link.turnPriority, newExtenderLength)
        } else {
            extender.resetStartsWith(to, link.turnPriority, newExtenderLength)
        }

        // update meets
        if (meets.startsWith == link.slideable) {
            meets.resetStartsWith(extender.slideable, meets.turnPriority, 0.0)
        } else {
            meets.resetEndsWith(extender.slideable, meets.turnPriority, 0.0)
        }
        fixSize(c, meets, meets.getLength(false), shape === TurnShape.G, false)
    }

    private fun logRectangularizationContext(vt4: VertexTurn, vt3: VertexTurn, vt2: VertexTurn, vt1: VertexTurn) {
        log.send(if (log.go()) null else "Context:")
        log.send(if (log.go()) null else "vt4 $vt4")
        log.send(if (log.go()) null else "vt3 $vt3")
        log.send(if (log.go()) null else "vt2 $vt2")
        log.send(if (log.go()) null else "vt1 $vt1")
    }

    /**
     * After the new darts are added to create the rectangle, this snips off the
     * old rectangle from the stack
     */
    private fun cutRectangleCorner(stack: MutableList<VertexTurn>, remove1: VertexTurn, remove2: VertexTurn) {
        stack.remove(remove1)
        stack.remove(remove2)
        log.send(if (log.go()) null else "Removed: $remove1")
        log.send(if (log.go()) null else "Removed: $remove2")
    }

    override val prefix: String
        get() = "ARec"
    override val isLoggingEnabled: Boolean
        get() = true

    companion object {
        /**
         * Works out the direction of turn between one segment and the next
         */
		@JvmStatic
		fun getTurn(t1: VertexTurn, t2: VertexTurn): Turn {
            val thisDirection = t1.direction
            val nextDirection = t2.direction
            return thisDirection.getDirectionChange(nextDirection)
        }

        fun <X> getIthElementRotating(items: List<X>, i: Int): X {
            return items[(i + items.size) % items.size]
        }

        @JvmStatic
		protected fun getItemRotating(stack: List<VertexTurn>, index: Int): VertexTurn {
            var index = index
            while (index < 0) index += stack.size
            index = index % stack.size
            return stack[index]
        }

        @JvmField
		var patternA = createList(Turn.LEFT, Turn.RIGHT, Turn.RIGHT)
        @JvmField
		var patternD = createList(Turn.RIGHT, Turn.RIGHT, Turn.LEFT)
        private fun createList(vararg turns: Turn): List<Turn> {
            val out: MutableList<Turn> = ArrayList()
            for (t in turns) {
                out.add(t)
            }
            return out
        }
    }
}